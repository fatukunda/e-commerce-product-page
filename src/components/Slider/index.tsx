import { useAppSelector, useAppDispatch } from "src/store/hooks";
import { IImage } from "src/models/Product";

import prevIcon from "../../assets/images/chevron-left.svg";
import nextIcon from "../../assets/images/chevron-right.svg";
import {
  getNextImage,
  getPreviousImage,
  setCurrentImage,
} from "src/store/slices/sliderSlice";

const Slider = () => {
  const images: IImage[] = useAppSelector((state) => state.slider.images);
  const currentImage: IImage = useAppSelector(
    (state) => state.slider.currentImage
  );
  const dispatch = useAppDispatch();

  return (
    <div className="relative lg:max-w-md">
      <button
        type="button"
        className="absolute left-4 top-[40%] flex aspect-square w-10 items-center justify-center rounded-full bg-white lg:-left-8 lg:top-[32%] lg:w-20"
        onClick={() => dispatch(getPreviousImage())}
        data-testid="get-previous-btn"
      >
        <img
          src={prevIcon}
          alt="previous"
          className="absolute top-1/2 left-1/2 h-fit max-w-full -translate-x-2 -translate-y-1/2"
        />
      </button>

      <img
        src={currentImage.src}
        alt="product"
        className="block w-full max-w-full lg:rounded-xl"
        data-testid="current-image"
      />

      <button
        type="button"
        className=" absolute right-4 top-[40%] flex aspect-square w-10 items-center justify-center rounded-full bg-white lg:-right-8 lg:top-[32%] lg:w-20"
        onClick={() => dispatch(getNextImage())}
        data-testid="get-next-btn"
      >
        <img
          src={nextIcon}
          alt="next"
          className="fit-w-full absolute top-1/2 left-1/2 h-fit -translate-x-1.5 -translate-y-1/2"
        />
      </button>

      <div className="hidden md:my-8 md:flex md:w-full md:items-center md:justify-evenly lg:justify-evenly">
        {images.map((image) => (
          <div
            key={image.id}
            className={`rounded-xl ${
              image.id === currentImage.id && "outline outline-primary"
            }`}
          >
            <img
              src={image.src}
              alt="product"
              className={`w-32 rounded-lg transition-all duration-300 hover:cursor-pointer hover:opacity-40 lg:w-[88px] ${
                image.id === currentImage.id && "opacity-40"
              }`}
              onClick={() => dispatch(setCurrentImage(image))}
              data-testid="selected-img"
            />
          </div>
        ))}
      </div>
    </div>
  );
};

export default Slider;
