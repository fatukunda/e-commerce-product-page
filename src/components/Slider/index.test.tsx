import { render, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider as ReduxProvider } from "react-redux";

import Slider from ".";
import mockStore from "../mocks/store";

const renderCart = () =>
  render(
    <ReduxProvider store={mockStore()}>
      <Slider />
    </ReduxProvider>
  );
describe("Tests for <Slider/>", () => {
  afterEach(cleanup);
  test("Slider: Snapshot", () => {
    const component = renderer.create(
      <ReduxProvider store={mockStore()}>
        <Slider />
      </ReduxProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders without crashing", () => {
    renderCart();
  });
  it("Displays the next image", () => {
    const { getByTestId } = renderCart();
    const getNextBtn = getByTestId("get-next-btn");
    const currentImage = getByTestId("current-image");

    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
    );
    fireEvent.click(getNextBtn);
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198"
    );
  });
  it("Displays the previouse image", () => {
    const { getByTestId } = renderCart();
    const getPrevBtn = getByTestId("get-previous-btn");
    const currentImage = getByTestId("current-image");
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
    );

    fireEvent.click(getPrevBtn);
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198"
    );
    fireEvent.click(getPrevBtn);
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
    );
  });
  it("Displays the selected image", () => {
    const { getAllByTestId, getByTestId } = renderCart();
    const imagesToSelect = getAllByTestId("selected-img");
    const currentImage = getByTestId("current-image");

    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
    );
    fireEvent.click(imagesToSelect[1]);
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198"
    );
    fireEvent.click(imagesToSelect[0]);
    expect(currentImage.getAttribute("src")).toBe(
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
    );
  });
});
