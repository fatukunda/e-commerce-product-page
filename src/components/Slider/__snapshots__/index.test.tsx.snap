// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Slider/> Slider: Snapshot 1`] = `
<div
  className="relative lg:max-w-md"
>
  <button
    className="absolute left-4 top-[40%] flex aspect-square w-10 items-center justify-center rounded-full bg-white lg:-left-8 lg:top-[32%] lg:w-20"
    data-testid="get-previous-btn"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="previous"
      className="absolute top-1/2 left-1/2 h-fit max-w-full -translate-x-2 -translate-y-1/2"
      src="chevron-left.svg"
    />
  </button>
  <img
    alt="product"
    className="block w-full max-w-full lg:rounded-xl"
    data-testid="current-image"
    src="https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
  />
  <button
    className=" absolute right-4 top-[40%] flex aspect-square w-10 items-center justify-center rounded-full bg-white lg:-right-8 lg:top-[32%] lg:w-20"
    data-testid="get-next-btn"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="next"
      className="fit-w-full absolute top-1/2 left-1/2 h-fit -translate-x-1.5 -translate-y-1/2"
      src="chevron-right.svg"
    />
  </button>
  <div
    className="hidden md:my-8 md:flex md:w-full md:items-center md:justify-evenly lg:justify-evenly"
  >
    <div
      className="rounded-xl outline outline-primary"
    >
      <img
        alt="product"
        className="w-32 rounded-lg transition-all duration-300 hover:cursor-pointer hover:opacity-40 lg:w-[88px] opacity-40"
        data-testid="selected-img"
        onClick={[Function]}
        src="https://cdn.shopify.com/s/files/1/0095/2274/1329/products/7_252F9_252Fe_252F2_252F79e29bdc358cf46c6f6df40f78ec1cabeda1a36a_21032_Top.jpg?v=1630500085"
      />
    </div>
    <div
      className="rounded-xl false"
    >
      <img
        alt="product"
        className="w-32 rounded-lg transition-all duration-300 hover:cursor-pointer hover:opacity-40 lg:w-[88px] false"
        data-testid="selected-img"
        onClick={[Function]}
        src="https://cdn.shopify.com/s/files/1/0095/2274/1329/products/a_252F2_252F4_252Ff_252Fa24f85e82f207a80c119cb64b0fa99d3ac4def70_21452.jpg?v=1630500198"
      />
    </div>
  </div>
</div>
`;
